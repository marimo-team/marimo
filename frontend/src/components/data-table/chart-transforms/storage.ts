/* Copyright 2024 Marimo. All rights reserved. */

import type { CellId } from "@/core/cells/ids";
import type { LineChartSchema } from "./chart-schemas";
import type { TypedString } from "@/utils/typed";
import { atomWithStorage } from "jotai/utils";
import type { z } from "zod";

export type TabName = TypedString<"TabName">;
const KEY = "marimo:charts";

export type ChartType = "line" | "bar" | "pie";

export interface TabStorage {
  cellId: CellId;
  tabName: TabName; // unique within cell
  chartType: ChartType;
  config: z.infer<typeof LineChartSchema>;
}

export const tabsStorageAtom = atomWithStorage<TabStorage[]>(KEY, []);
