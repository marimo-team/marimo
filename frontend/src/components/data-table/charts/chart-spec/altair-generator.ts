/* Copyright 2024 Marimo. All rights reserved. */
import {
  FunctionCall,
  Literal,
  type PythonCode,
  Variable,
  VariableDeclaration,
} from "@/utils/python-poet/poet";

import type { VegaLiteSpec } from "@/plugins/impl/vega/types";

/**
 * Generates Python code for an Altair chart.
 */
export function generateAltairChart(
  spec: VegaLiteSpec,
  datasource: string,
): PythonCode {
  let code = new FunctionCall("alt.Chart", [new Variable(datasource)], true);

  const hasMarkProperty = "mark" in spec;
  if (hasMarkProperty) {
    const markType =
      typeof spec.mark === "string" ? spec.mark : spec.mark?.type;

    if (markType) {
      let markProps: Record<string, PythonCode> = {};

      if (typeof spec.mark === "object" && "type" in spec.mark) {
        markProps = Object.fromEntries(
          Object.entries(spec.mark)
            .filter(([key]) => key !== "type")
            .map(([key, value]) => [key, new Literal(value)]),
        );
      }

      code = code.chain(`mark_${markType}`, markProps);
    }
  }

  const hasEncodings = "encoding" in spec;
  if (hasEncodings) {
    const encodeArgs: Record<string, PythonCode> = {};
    const encodings = spec.encoding;

    if (encodings?.x) {
      const kwargs = makeKwargs(encodings.x);
      encodeArgs.x = new FunctionCall("alt.X", kwargs);
    }

    if (encodings?.y) {
      const kwargs = makeKwargs(encodings.y);
      encodeArgs.y = new FunctionCall("alt.Y", kwargs);
    }

    if (encodings?.color) {
      const kwargs = makeKwargs(encodings.color);
      encodeArgs.color = new FunctionCall("alt.Color", kwargs);
    }

    if (encodings?.theta) {
      const kwargs = makeKwargs(encodings.theta);
      encodeArgs.theta = new FunctionCall("alt.Theta", kwargs);
    }

    const hasRow = encodings && "row" in encodings;
    if (hasRow && encodings.row) {
      const kwargs = makeKwargs(encodings.row);
      encodeArgs.row = new FunctionCall("alt.Row", kwargs);
    }

    const hasColumn = encodings && "column" in encodings;
    if (hasColumn && encodings.column) {
      const kwargs = makeKwargs(encodings.column);
      encodeArgs.column = new FunctionCall("alt.Column", kwargs);
    }

    if (encodings?.tooltip) {
      const tooltip = encodings.tooltip;
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const makeTooltip = (t: Record<string, any>) => {
        const kwargs = makeKwargs(t);
        return new FunctionCall("alt.Tooltip", kwargs);
      };

      const tooltipArray = Array.isArray(tooltip)
        ? new Literal(tooltip.map(makeTooltip))
        : makeTooltip(tooltip);
      encodeArgs.tooltip = tooltipArray;
    }

    code = code.chain("encode", encodeArgs);
  }

  if (spec.resolve) {
    const resolve = spec.resolve;
    const axisArgs: Record<string, PythonCode> = {};

    if (resolve.axis?.x) {
      axisArgs.x = new Literal(resolve.axis.x);
    }

    if (resolve.axis?.y) {
      axisArgs.y = new Literal(resolve.axis.y);
    }

    code = code.chain("resolve_scale", axisArgs);
  }

  const propertiesArgs: Record<string, PythonCode> = {};
  const propertiesKeys = ["title", "height", "width"];
  for (const key of propertiesKeys) {
    if (key in spec) {
      propertiesArgs[key] = new Literal(spec[key as keyof VegaLiteSpec]);
    }
  }

  if (Object.keys(propertiesArgs).length > 0) {
    code = code.chain("properties", propertiesArgs);
  }

  return code;
}

/**
 * Wraps altair code in a variable declaration.
 *
 * @param chartType - The type of chart to generate.
 * @param spec - The specification of the chart.
 * @param datasource - The name of the datasource to use.
 * @param variableName - The name of the variable to declare.
 */
export function generateAltairChartSnippet(
  spec: VegaLiteSpec,
  datasource: string,
  variableName: string,
): string {
  const code = generateAltairChart(spec, datasource).toCode();
  return `
${new VariableDeclaration(variableName, code).toCode()}
${variableName}
  `.trim();
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function makeKwargs<T extends Record<string, any>>(obj: T) {
  const result: Record<string, PythonCode> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (value !== undefined) {
      result[key] = new Literal(value);
    }
  }

  return result;
}
