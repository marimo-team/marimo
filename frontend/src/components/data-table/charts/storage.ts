/* Copyright 2024 Marimo. All rights reserved. */

import type { CellId } from "@/core/cells/ids";
import type { TypedString } from "@/utils/typed";
import { atomWithStorage } from "jotai/utils";
import { z } from "zod";
import { ChartSchema } from "./schemas";
import { Logger } from "@/utils/Logger";
import type { ChartType } from "./types";
import { NotebookScopedLocalStorage } from "@/utils/localStorage";
import { capitalize } from "lodash-es";

export type TabName = TypedString<"TabName">;
export const KEY = "marimo:charts:v2";

interface TabStorage {
  tabName: TabName; // unique within cell
  chartType: ChartType;
  config: z.infer<typeof ChartSchema>;
}

type TabStorageMap = Map<CellId, TabStorage[]>;

const TabStorageSchema = z.object({
  tabName: z.string().transform((name) => name as TabName),
  chartType: z.string().transform((type) => type as ChartType),
  config: ChartSchema,
});
const TabStorageEntriesSchema = z.array(
  z.tuple([
    z.string().transform((name) => name as CellId),
    z.array(TabStorageSchema),
  ]),
);

const storage = new NotebookScopedLocalStorage(
  KEY,
  TabStorageEntriesSchema,
  () => [],
);

// Custom storage adapter to ensure objects are serialized as maps
const mapStorage = {
  getItem: (key: string): TabStorageMap => {
    try {
      const value = storage.get(key);
      return new Map(value);
    } catch (error) {
      Logger.warn("Error getting chart storage", error);
      return new Map();
    }
  },
  setItem: (key: string, value: TabStorageMap): void => {
    storage.set(key, [...value.entries()]);
  },
  removeItem: (key: string): void => {
    storage.remove(key);
  },
};

export const tabsStorageAtom = atomWithStorage<TabStorageMap>(
  KEY,
  new Map(),
  mapStorage,
);

/**
 * Convenience function to get the tab name for a given tab number and chart type
 */
export function getChartTabName(tabNum: number, chartType: ChartType) {
  const capitalizedChartType = capitalize(chartType);
  if (tabNum === 0) {
    return `${capitalizedChartType} Chart` as TabName;
  }
  return `${capitalizedChartType} Chart ${tabNum + 1}` as TabName;
}

const initialStorageValue = mapStorage.getItem(KEY);
/**
 * Returns true if the cell has a chart
 */
export function hasChart(cellId: CellId) {
  const tabs = initialStorageValue.get(cellId);
  if (!tabs) {
    return false;
  }
  return tabs.length > 0;
}
