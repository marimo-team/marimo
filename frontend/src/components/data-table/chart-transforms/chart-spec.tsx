/* Copyright 2024 Marimo. All rights reserved. */

import type { TopLevelSpec } from "vega-lite";
import { ChartType } from "./storage";
import type { ResolvedTheme } from "@/theme/useTheme";
import type { DataType } from "@/core/kernel/messages";
import {
  type BinSchema,
  DEFAULT_AGGREGATION,
  DEFAULT_BIN_VALUE,
  NONE_GROUP_BY,
  type ChartSchema,
  DEFAULT_COLOR_SCHEME,
} from "./chart-schemas";
import type { z } from "zod";
import type { Mark } from "@/plugins/impl/vega/types";
import { logNever } from "@/utils/assertNever";
import type { Type } from "vega-lite/build/src/type";
import type {
  ColorDef,
  OffsetDef,
  PolarDef,
  PositionDef,
  StringFieldDef,
} from "vega-lite/build/src/channeldef";
import type { ColorScheme } from "vega";

export function createVegaSpec(
  chartType: ChartType,
  data: object[],
  formValues: z.infer<typeof ChartSchema>,
  theme: ResolvedTheme,
  width: number | "container",
  height: number,
): TopLevelSpec | null {
  let xAxisLabel = formValues.general.xColumn?.field;
  let yAxisLabel = formValues.general.yColumn?.field;

  if (
    formValues.general.yColumn?.agg &&
    formValues.general.yColumn.agg !== DEFAULT_AGGREGATION
  ) {
    yAxisLabel = `${formValues.general.yColumn.agg.toUpperCase()}(${yAxisLabel})`;
  }

  if (formValues.xAxis?.label && formValues.xAxis.label.trim() !== "") {
    xAxisLabel = formValues.xAxis.label;
  }

  if (formValues.yAxis?.label && formValues.yAxis.label.trim() !== "") {
    yAxisLabel = formValues.yAxis.label;
  }

  const xEncodingKey = chartType === ChartType.PIE ? "theta" : "x";
  const yEncodingKey = chartType === ChartType.PIE ? "color" : "y";

  const groupByFieldExists =
    formValues.general.groupByColumn?.field !== NONE_GROUP_BY;

  const shouldApplyStackingToX =
    groupByFieldExists && formValues.general.horizontal;
  const shouldApplyStackingToY =
    groupByFieldExists && !formValues.general.horizontal;

  const xEncoding: PositionDef<string> | PolarDef<string> = {
    field: formValues.general.xColumn?.field,
    type: convertDataTypeToVegaType(
      formValues.general.xColumn?.type ?? "unknown",
    ),
    bin: getBin(formValues.xAxis?.bin),
    title: xAxisLabel,
    stack: shouldApplyStackingToX ? formValues.general.stacking : undefined,
  };

  const colorInScale = getColorInScale(formValues);

  const yEncoding: PositionDef<string> | PolarDef<string> = {
    field: formValues.general.yColumn?.field,
    type: convertDataTypeToVegaType(
      formValues.general.yColumn?.type ?? "unknown",
    ),
    bin: getBin(formValues.yAxis?.bin),
    title: yAxisLabel,
    // If color encoding is used as y, we can define the scheme here
    scale:
      colorInScale && yEncodingKey === "color"
        ? { ...colorInScale }
        : undefined,
    stack: shouldApplyStackingToY ? formValues.general.stacking : undefined,
  };

  const schema: TopLevelSpec = {
    $schema: "https://vega.github.io/schema/vega-lite/v5.json",
    background: theme === "dark" ? "dark" : "white",
    title: formValues.general.title,
    data: {
      values: data,
    },
    height: height,
    width: width,
    mark: {
      type: convertChartTypeToMark(chartType),
    },
    encoding: {
      [xEncodingKey]: formValues.general.horizontal ? yEncoding : xEncoding,
      [yEncodingKey]: formValues.general.horizontal ? xEncoding : yEncoding,
      xOffset: getOffset(chartType, formValues),
      ...getColor(chartType, formValues),
      tooltip: getTooltips(formValues),
    },
  };
  return schema;
}

// color can be used for grouping
// it can also conflict with the color (y) encoding for pie charts, so we return undefined
function getColor(
  chartType: ChartType,
  formValues: z.infer<typeof ChartSchema>,
) {
  if (
    chartType === ChartType.PIE ||
    formValues.general.groupByColumn?.field === NONE_GROUP_BY
  ) {
    return undefined;
  }

  const colorDef: ColorDef<string> = {
    field: formValues.general.groupByColumn?.field,
    type: convertDataTypeToVegaType(
      formValues.general.groupByColumn?.type ?? "unknown",
    ),
    scale: {
      ...getColorInScale(formValues),
    },
  };

  return {
    color: colorDef,
  };
}

function getColorInScale(formValues: z.infer<typeof ChartSchema>) {
  const colorRange = formValues.color?.range;
  if (colorRange && colorRange.length > 0) {
    return {
      range: colorRange,
    };
  }

  const scheme = formValues.color?.scheme;
  if (scheme === DEFAULT_COLOR_SCHEME) {
    return undefined;
  }
  return {
    scheme: scheme as ColorScheme,
  };
}

function getOffset(
  chartType: ChartType,
  formValues: z.infer<typeof ChartSchema>,
): OffsetDef<string> | undefined {
  if (
    formValues.general.stacking ||
    formValues.general.groupByColumn?.field === NONE_GROUP_BY ||
    chartType === ChartType.PIE
  ) {
    return undefined;
  }
  return {
    field:
      formValues.general.groupByColumn?.field === NONE_GROUP_BY
        ? undefined
        : formValues.general.groupByColumn?.field,
  };
}

function getBin(binValues?: z.infer<typeof BinSchema>) {
  if (binValues?.binned) {
    if (binValues.step === DEFAULT_BIN_VALUE) {
      return true;
    }

    return {
      binned: true,
      step: binValues.step,
    };
  }
}

function getTooltips(formValues: z.infer<typeof ChartSchema>) {
  return formValues.general.tooltips?.map(
    (tooltip): StringFieldDef<string> => ({
      field: tooltip.field,
      aggregate: (() => {
        if (tooltip.field !== formValues.general.yColumn?.field) {
          return undefined;
        }
        return formValues.general.yColumn?.agg === DEFAULT_AGGREGATION
          ? undefined
          : formValues.general.yColumn?.agg;
      })(),
      format: getTooltipFormat(tooltip.type),
    }),
  );
}

function getTooltipFormat(dataType: DataType): string | undefined {
  switch (dataType) {
    case "integer":
      return ",d";
    case "number":
      return ".2f";
    default:
      return undefined;
  }
}

function convertDataTypeToVegaType(dataType: DataType): Type {
  switch (dataType) {
    case "number":
    case "integer":
      return "quantitative";
    case "string":
      return "nominal";
    case "boolean":
      return "nominal";
    case "date":
    case "datetime":
    case "time":
      return "temporal";
    case "unknown":
      return "nominal";
    default:
      logNever(dataType);
      return "nominal";
  }
}

function convertChartTypeToMark(chartType: ChartType): Mark {
  switch (chartType) {
    case ChartType.PIE:
      return "arc";
    case ChartType.SCATTER:
      return "point";
    default:
      return chartType;
  }
}
