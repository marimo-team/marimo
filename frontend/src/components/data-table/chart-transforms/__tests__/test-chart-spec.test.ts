/* Copyright 2024 Marimo. All rights reserved. */

import { describe, it, expect } from "vitest";
import { createVegaSpec } from "../chart-spec";
import { ChartType } from "../storage";
import {
  DEFAULT_AGGREGATION,
  DEFAULT_BIN_VALUE,
  NONE_GROUP_BY,
} from "../chart-schemas";
import type { z } from "zod";
import type { ChartSchema } from "../chart-schemas";

describe("createVegaSpec", () => {
  // Sample data for testing
  const sampleData = [
    { category: "A", value: 10, group: "Group 1" },
    { category: "B", value: 20, group: "Group 1" },
    { category: "C", value: 15, group: "Group 1" },
    { category: "A", value: 5, group: "Group 2" },
    { category: "B", value: 10, group: "Group 2" },
    { category: "C", value: 25, group: "Group 2" },
  ];

  // Common test parameters
  const width = 400;
  const height = 300;

  // Helper function to create basic form values
  const createBasicFormValues = (): z.infer<typeof ChartSchema> => ({
    general: {
      title: "Test Chart",
      xColumn: {
        field: "category",
        type: "string" as const,
      },
      yColumn: {
        field: "value",
        type: "number" as const,
        agg: DEFAULT_AGGREGATION as "default",
      },
    },
  });

  describe("Bar Chart", () => {
    it("should create a horizontal bar chart", () => {
      const formValues = {
        ...createBasicFormValues(),
        general: {
          ...createBasicFormValues().general,
          horizontal: true,
        },
      };

      const spec = createVegaSpec(
        ChartType.BAR,
        sampleData,
        formValues,
        "light",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });

    it("should create a stacked bar chart with grouping", () => {
      const formValues = {
        ...createBasicFormValues(),
        general: {
          ...createBasicFormValues().general,
          groupByColumn: {
            field: "group",
            type: "string" as const,
          },
          stacking: true,
        },
      };

      const spec = createVegaSpec(
        ChartType.BAR,
        sampleData,
        formValues,
        "light",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });
  });

  it("should create a bar chart with binning", () => {
    const formValues = {
      ...createBasicFormValues(),
      xAxis: { bin: { binned: true, step: DEFAULT_BIN_VALUE } },
    };

    const spec = createVegaSpec(
      ChartType.BAR,
      sampleData,
      formValues,
      "light",
      width,
      height,
    );

    expect(spec).toMatchSnapshot();
  });

  describe("Line Chart", () => {
    it("should create a basic line chart spec", () => {
      const formValues = createBasicFormValues();
      const spec = createVegaSpec(
        ChartType.LINE,
        sampleData,
        formValues,
        "light",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });
  });

  describe("Pie Chart", () => {
    it("should create a pie chart with tooltips", () => {
      const formValues = {
        ...createBasicFormValues(),
        general: {
          ...createBasicFormValues().general,
          tooltips: [
            { field: "category", type: "string" as const },
            { field: "value", type: "number" as const },
          ],
        },
      };

      const spec = createVegaSpec(
        ChartType.PIE,
        sampleData,
        formValues,
        "light",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });
  });

  describe("Scatter Chart", () => {
    it("should create a scatter chart with grouping", () => {
      const formValues = {
        ...createBasicFormValues(),
        general: {
          ...createBasicFormValues().general,
          groupByColumn: {
            field: "group",
            type: "string" as const,
          },
        },
      };

      const spec = createVegaSpec(
        ChartType.SCATTER,
        sampleData,
        formValues,
        "light",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });
  });

  describe("Theme variations", () => {
    it("should create a chart with dark theme", () => {
      const formValues = createBasicFormValues();
      const spec = createVegaSpec(
        ChartType.BAR,
        sampleData,
        formValues,
        "dark",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });
  });

  describe("Edge cases", () => {
    it("should handle missing xColumn field", () => {
      const formValues = {
        ...createBasicFormValues(),
        general: {
          ...createBasicFormValues().general,
          xColumn: {
            field: undefined,
            type: "string" as const,
          },
        },
      };

      const spec = createVegaSpec(
        ChartType.BAR,
        sampleData,
        formValues,
        "light",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });

    it("should handle missing yColumn field", () => {
      const formValues = {
        ...createBasicFormValues(),
        general: {
          ...createBasicFormValues().general,
          yColumn: {
            field: undefined,
            type: "number" as const,
            agg: DEFAULT_AGGREGATION as "default",
          },
        },
      };

      const spec = createVegaSpec(
        ChartType.BAR,
        sampleData,
        formValues,
        "light",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });

    it("should handle NONE_GROUP_BY for groupByColumn", () => {
      const formValues = {
        ...createBasicFormValues(),
        general: {
          ...createBasicFormValues().general,
          groupByColumn: {
            field: NONE_GROUP_BY,
            type: "string" as const,
          },
          stacking: true,
        },
      };

      const spec = createVegaSpec(
        ChartType.BAR,
        sampleData,
        formValues,
        "light",
        width,
        height,
      );

      expect(spec).toMatchSnapshot();
    });
  });
});
