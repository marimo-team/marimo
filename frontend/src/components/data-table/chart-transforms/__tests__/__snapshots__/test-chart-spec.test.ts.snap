// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createVegaSpec > Bar Chart > should create a horizontal bar chart 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": undefined,
      "scale": {
        "scheme": undefined,
      },
      "type": "nominal",
    },
    "tooltip": undefined,
    "x": {
      "bin": undefined,
      "field": "value",
      "scale": undefined,
      "stack": undefined,
      "title": "value",
      "type": "quantitative",
    },
    "xOffset": {
      "field": undefined,
    },
    "y": {
      "bin": undefined,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Bar Chart > should create a stacked bar chart with grouping 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": "group",
      "scale": {
        "scheme": undefined,
      },
      "type": "nominal",
    },
    "tooltip": undefined,
    "x": {
      "bin": undefined,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
    "xOffset": undefined,
    "y": {
      "bin": undefined,
      "field": "value",
      "scale": undefined,
      "stack": true,
      "title": "value",
      "type": "quantitative",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Edge cases > should handle NONE_GROUP_BY for groupByColumn 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "tooltip": undefined,
    "x": {
      "bin": undefined,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
    "xOffset": undefined,
    "y": {
      "bin": undefined,
      "field": "value",
      "scale": undefined,
      "stack": undefined,
      "title": "value",
      "type": "quantitative",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Edge cases > should handle missing xColumn field 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": undefined,
      "scale": {
        "scheme": undefined,
      },
      "type": "nominal",
    },
    "tooltip": undefined,
    "x": {
      "bin": undefined,
      "field": undefined,
      "stack": undefined,
      "title": undefined,
      "type": "nominal",
    },
    "xOffset": {
      "field": undefined,
    },
    "y": {
      "bin": undefined,
      "field": "value",
      "scale": undefined,
      "stack": undefined,
      "title": "value",
      "type": "quantitative",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Edge cases > should handle missing yColumn field 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": undefined,
      "scale": {
        "scheme": undefined,
      },
      "type": "nominal",
    },
    "tooltip": undefined,
    "x": {
      "bin": undefined,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
    "xOffset": {
      "field": undefined,
    },
    "y": {
      "bin": undefined,
      "field": undefined,
      "scale": undefined,
      "stack": undefined,
      "title": undefined,
      "type": "quantitative",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Line Chart > should create a basic line chart spec 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": undefined,
      "scale": {
        "scheme": undefined,
      },
      "type": "nominal",
    },
    "tooltip": undefined,
    "x": {
      "bin": undefined,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
    "xOffset": {
      "field": undefined,
    },
    "y": {
      "bin": undefined,
      "field": "value",
      "scale": undefined,
      "stack": undefined,
      "title": "value",
      "type": "quantitative",
    },
  },
  "height": 300,
  "mark": {
    "type": "line",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Pie Chart > should create a pie chart with tooltips 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "bin": undefined,
      "field": "value",
      "scale": {
        "scheme": undefined,
      },
      "stack": undefined,
      "title": "value",
      "type": "quantitative",
    },
    "theta": {
      "bin": undefined,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
    "tooltip": [
      {
        "aggregate": undefined,
        "field": "category",
        "format": undefined,
      },
      {
        "aggregate": undefined,
        "field": "value",
        "format": ".2f",
      },
    ],
    "xOffset": undefined,
  },
  "height": 300,
  "mark": {
    "type": "arc",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Scatter Chart > should create a scatter chart with grouping 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": "group",
      "scale": {
        "scheme": undefined,
      },
      "type": "nominal",
    },
    "tooltip": undefined,
    "x": {
      "bin": undefined,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
    "xOffset": {
      "field": "group",
    },
    "y": {
      "bin": undefined,
      "field": "value",
      "scale": undefined,
      "stack": undefined,
      "title": "value",
      "type": "quantitative",
    },
  },
  "height": 300,
  "mark": {
    "type": "point",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Theme variations > should create a chart with dark theme 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "dark",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": undefined,
      "scale": {
        "scheme": undefined,
      },
      "type": "nominal",
    },
    "tooltip": undefined,
    "x": {
      "bin": undefined,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
    "xOffset": {
      "field": undefined,
    },
    "y": {
      "bin": undefined,
      "field": "value",
      "scale": undefined,
      "stack": undefined,
      "title": "value",
      "type": "quantitative",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > should create a bar chart with binning 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": undefined,
      "scale": {
        "scheme": undefined,
      },
      "type": "nominal",
    },
    "tooltip": undefined,
    "x": {
      "bin": true,
      "field": "category",
      "stack": undefined,
      "title": "category",
      "type": "nominal",
    },
    "xOffset": {
      "field": undefined,
    },
    "y": {
      "bin": undefined,
      "field": "value",
      "scale": undefined,
      "stack": undefined,
      "title": "value",
      "type": "quantitative",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "title": "Test Chart",
  "width": 400,
}
`;
